### PR TITLE
when a user was deleted remove them from applicable list, unless...

### LINKED ISSUE #15258: External Storage - WebDAV - New user has access to storage from deleted user (same user-ID)
### Steps to reproduce
1. Delete existing user account "user3"
2. Create new user  account "user3"
3.

### Expected behaviour
New User "user3" should not have acces to external webdav storage configured by the deleted user "user3".
The Config in database shoul be deletet

### Actual behaviour
Newly created user with same user-id of a deleted user has access to the configured external webdav storage from the deleted user

### Server configuration

**Operating system:** openSUSE Leap 15.0

**Web server:** Apache 2.4.33

**Database:** MariaDB 10.2.22

**PHP version:** 7.2

**Nextcloud version:** 16.0.0.9

**Updated from an older Nextcloud/ownCloud or fresh install:** update from latest 15er

**Where did you install Nextcloud from:** opensuse repository php:applications

**Signing status:** No errors have been found.

**List of activated apps:**
<details>
<summary>
Enabled:
  - accessibility: 1.2.0
  - activity: 2.9.1
  - bruteforcesettings: 1.3.0
  - calendar: 1.7.0
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - contacts: 3.1.1
  - dav: 1.9.2
  - drop_account: 0.1.0
  - encryption: 2.4.0
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_external: 1.7.0
  - files_pdfviewer: 1.5.0
  - files_rightclick: 0.13.0
  - files_sharing: 1.8.0
  - files_texteditor: 2.8.0
  - files_trashbin: 1.6.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - gallery: 18.3.0
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - nextcloud_announcements: 1.5.0
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - password_policy: 1.6.0
  - polls: 0.10.2
  - privacy: 1.0.0
  - provisioning_api: 1.6.0
  - recommendations: 0.4.0
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - support: 1.0.0
  - survey_client: 1.4.0
  - systemtags: 1.6.0
  - tasks: 0.9.8
  - theming: 1.7.0
  - twofactor_backupcodes: 1.5.0
  - twofactor_totp: 2.1.2
  - unsplash: 1.1.3
  - updatenotification: 1.6.0
  - viewer: 1.0.0
  - workflowengine: 1.6.0
Disabled:
  - admin_audit
  - twofactor_admin
  - user_ldap

</summary>
</details>

**Nextcloud configuration:**
<details>
<summary>
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "127.0.0.1",
            "192.168.0.206",
            "192.168.0.241",
            "*"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.0.9",
        "overwrite.cli.url": "http:\/\/localhost\/nextcloud",
        "htaccess.RewriteBase": "\/nextcloud",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "maintenance": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "default_language": "de",
        "default_locale": "de_DE",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "tls",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_sendmailmode": "smtp",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "production",
        "loglevel": 2,
        "data-fingerprint": "dbaf646e0bfe174824f11fa27050b695",
        "app_install_overwrite": [
            "tasks"
        ]
    }
}

</summary>
</details>

**Are you using external storage, if yes which one:** WebDAV

**Are you using encryption:** yes

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Latest Firefox, Chrome, Brave

**Operating system:** Linux (openSUSE Leap 15), Windows 10

#### Web server error log
<details>
no errors relevant for this bugreport
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
no errors relevant for this bugreport

![oc_external_mounts](https://user-images.githubusercontent.com/38841410/56848554-c5b32a00-68ea-11e9-9c51-d9a6bc239e5c.png)
![oc_external_config](https://user-images.githubusercontent.com/38841410/56848556-d794cd00-68ea-11e9-87dc-ad3a9176c9a9.png)
![user3_external-storage-config](https://user-images.githubusercontent.com/38841410/56848570-f85d2280-68ea-11e9-95ea-4204f7e47d08.png)

```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
no errors relevant for this bugreport
```
</details>

### PR DESCRIPTION
**Pull request by blizzz**
it is the only user, then remove the mount

fixes #15258

@icewind1991 does it make sense to you? 

* [x] Then, we need basically the same for groups.

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]

File: apps/files_external/lib/Service/DBConfigService.php [hunk 1]
```diff
@@ -114,6 +114,30 @@ public function getMountsForUser($userId, $groupIds) {
 		return $this->getMountsFromQuery($query);
 	}
 
+	public function modifyMountsOnUserDelete(string $uid) {
+		$builder = $this->connection->getQueryBuilder();
+		$query = $builder->select(['a.mount_id', $builder->createFunction('COUNT(a.mount_id) as count')])
```
Comment (nickvergessen) [inline apps/files_external/lib/Service/DBConfigService.php]: Please use `$builder->func()->count('a.mount_id')`

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by blizzz** [comment]
Ah, edge case. It will remove mounts who are made accessible for groups plus this user.

File: apps/files_external/lib/Service/DBConfigService.php [hunk 2]
```diff
@@ -114,6 +114,38 @@ public function getMountsForUser($userId, $groupIds) {
 		return $this->getMountsFromQuery($query);
 	}
 
+	public function modifyMountsOnUserDelete(string $uid) {
+		$this->modifyMountsOnDelete($uid, self::APPLICABLE_TYPE_GROUP);
```
Comment (nickvergessen) [inline apps/files_external/lib/Service/DBConfigService.php]: Möp

File: apps/files_external/lib/Service/DBConfigService.php [hunk 3]
```diff
@@ -114,6 +114,38 @@ public function getMountsForUser($userId, $groupIds) {
 		return $this->getMountsFromQuery($query);
 	}
 
+	public function modifyMountsOnUserDelete(string $uid) {
```
Comment (nickvergessen) [inline apps/files_external/lib/Service/DBConfigService.php]: Vooid

Comment (nickvergessen) [reply apps/files_external/lib/Service/DBConfigService.php]: Same on all other methods

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/files_external/lib/Service/DBConfigService.php (hunk 2 above)
Comment (blizzz) [reply apps/files_external/lib/Service/DBConfigService.php]: only checking whether you are paying attention 😇

**Comment by blizzz** [comment]
@nickvergessen @icewind1991 can we get this in?

**Review by icewind1991 — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

File: apps/files_external/lib/Service/DBConfigService.php [hunk 4]
```diff
@@ -114,6 +114,38 @@ public function getMountsForUser($userId, $groupIds) {
...
+
+	public function modifyMountsOnGroupDelete(string $gid): void {
+		$this->modifyMountsOnDelete($gid, self::APPLICABLE_TYPE_GROUP);
+	}
+
+	protected function modifyMountsOnDelete(string $applicableId, int $applicableType): void {
+		$builder = $this->connection->getQueryBuilder();
+		$query = $builder->select(['a.mount_id', $builder->func()->count('a.mount_id', 'count')])
+			->from('external_applicable', 'a')
+			->rightJoin('a', 'external_applicable', 'b', $builder->expr()->eq('a.mount_id', 'b.mount_id'))
```
Comment (juliushaertl) [inline apps/files_external/lib/Service/DBConfigService.php]: Right join was not supported on sqlite iirc

**Comment by juliushaertl** [comment]
Breaks quite some integration tests: https://drone.nextcloud.com/nextcloud/server/23756/32/3

File: apps/files_external/lib/Service/DBConfigService.php (hunk 4 above)
Comment (blizzz) [reply apps/files_external/lib/Service/DBConfigService.php]: turning it around seems to work, same table after all. PR incoming,

**Pull request merged**